### PR TITLE
[Feature] Support partition prune for range partition by expr table

### PR DESCRIPTION
**Pull request by Astralidea**
## Problem Summary:
This is the follow-up to this PR #23059
The main solution is to hit the corresponding partition when querying.

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 95 (27.37%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|0|29|00.00%|[234, 235, 236, 237, 241, 242, 243, 244, 245, 247, 248, 253, 254, 255, 256, 257, 258, 259, 260, 264, 269, 270, 271, 273, 278, 279, 280, 281, 283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ColumnFilterConverter.java|1|38|02.63%|[90, 91, 94, 95, 99, 101, 102, 105, 106, 115, 116, 138, 139, 141, 142, 145, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 158, 159, 160, 162, 163, 165, 166, 167, 168, 169, 171]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|4|6|66.67%|[327, 328]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptDistributionPruner.java|4|5|80.00%|[66]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/PartitionColPredicateExtractor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|13|13|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 72 / 95 (75.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ColumnFilterConverter.java|21|38|55.26%|[90, 91, 94, 95, 99, 101, 102, 105, 106, 115, 116, 139, 151, 156, 160, 166, 167]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|4|6|66.67%|[327, 328]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptDistributionPruner.java|4|5|80.00%|[66]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|26|29|89.66%|[264, 271, 283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/PartitionColPredicateExtractor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|13|13|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/ColumnFilterConverter.java [hunk 1]
```diff
@@ -81,12 +85,37 @@ public static Map<String, PartitionColumnFilter> convertColumnFilter(List<Scalar
...
+        }
+
+        if (!checkColumnRefCanPartition(predicate.getChild(0), table)) {
+            return;
+        }
+
+        // rewrite invalid date cast expr to NullLiteral
+        ScalarOperator rewritePredicate = rewriteInvalidDateCast(predicate);
+
+        if (rewritePredicate.getChildren().stream().skip(1).anyMatch(d -> !OperatorType.CONSTANT.equals(d.getOpType()))) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/ColumnFilterConverter.java:101]: rewritePredicate.isConstant()

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 74 / 96 (77.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ColumnFilterConverter.java|21|38|55.26%|[90, 91, 94, 95, 99, 101, 102, 105, 106, 115, 116, 139, 151, 156, 160, 166, 167]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|4|6|66.67%|[327, 328]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptDistributionPruner.java|4|5|80.00%|[66]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|28|30|93.33%|[268, 275]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/PartitionColPredicateExtractor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|13|13|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java [hunk 2]
```diff
@@ -218,6 +225,64 @@ public static Map<String, Database> collectAllDatabase(ConnectContext context, S
         return dbs;
     }
 
+    // replaces a field in an expression with a constant
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java:228]: best move to ColumnFilterConverter.java?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 74 / 96 (77.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|4|6|66.67%|[327, 328]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ColumnFilterConverter.java|43|62|69.35%|[113, 120, 141, 142, 145, 146, 150, 152, 153, 156, 157, 166, 167, 190, 202, 207, 211, 217, 218]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptDistributionPruner.java|4|5|80.00%|[66]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/PartitionColPredicateExtractor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|13|13|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**